File: benchmarks/84-direct-testbed/direct_testbed.m
(* Experiment 84 -- DIRECT across the NMinimize test corpus.

   Runs the RACEABLE (continuous, box-bounded) NMinimize stress / unit-test
   problems from tests/test_nminimize.c through the DIRECT engine
   (Method -> {"DIRECT", ...}) and compares each to scipy.optimize.direct on the
   identical box. DIRECT is deterministic -- no seed. Both sides use scipy's
   default parameters except where a case is annotated LF/N in its label
   ("LocallyBiased" -> False, "MaxFunctionEvaluations" -> N) -- the same
   appropriate parameterisation on both sides. The race is raw-vs-raw --
   scipy's direct does no local polish, so "PostProcess" -> False disables
   Mathilda's.

   Every case reaches the SAME point in both systems: the global for most, a
   shared non-global basin for Rosenbrock-10D (~8.7916) and Gaussian-well-10D
   (both raw DIRECTs stall at 1.0 -- the well at 1.2345 is too narrow for
   center-based subdivision). Schwefel-10D is excluded (the two deterministic
   subdivisions diverge into different missed basins).

   The rest of the corpus is analysed in README.md and is NOT a race:
   scipy.optimize.direct is BOX-ONLY, so the constrained / equality /
   disjunctive / integer problems have no scipy column (Mathilda's DIRECT solves
   the low-dimensional ones via its penalty + augmented-Lagrangian polish, which
   scipy's direct cannot do); the high-dimensional equality-constrained A-series
   and the combinatorial B-series are beyond DIRECT in either system.

   The check is the objective at the reported point, rounded per problem so both
   implementations agree at the basin. Multi-variable objectives are built over
   fresh global symbols and passed through Evaluate[] past NMinimize's argument
   handling (so they compile to the machine path). The accumulator globals are
   named so they can never collide with a problem variable Symbol[prefix <> i]. *)

Get["../harness.m"];

require[{"NMinimize"}];

R  = {"DIRECT", "PostProcess" -> False};                                  (* default params *)
Regg = {"DIRECT", "PostProcess" -> False, "LocallyBiased" -> False, "MaxFunctionEvaluations" -> 20000};
Rgw  = {"DIRECT", "PostProcess" -> False, "LocallyBiased" -> False, "MaxFunctionEvaluations" -> 40000};

(* Fresh-symbol objective / variable / box triples, precomputed out of the
   timed region (the 79-shgo idiom). *)
griObj[n_] := With[{v = Table[Symbol["gr" <> ToString[i]], {i, 1, n}]},
    Total[Table[v[[i]]^2/4000, {i, 1, n}]] - Product[Cos[v[[i]]/Sqrt[i]], {i, 1, n}] + 1];
griVar[n_] := Table[Symbol["gr" <> ToString[i]], {i, 1, n}];
griBox[n_, w_] := And @@ Table[-w <= Symbol["gr" <> ToString[i]] <= w, {i, 1, n}];

sqObj[n_] := With[{v = Table[Symbol["sq" <> ToString[i]], {i, 1, n}]},
    Total[Table[10^(6 (i - 1)/9) v[[i]]^2, {i, 1, n}]]];
sqVar[n_] := Table[Symbol["sq" <> ToString[i]], {i, 1, n}];
sqBox[n_] := And @@ Table[-10 <= Symbol["sq" <> ToString[i]] <= 10, {i, 1, n}];

qbObj[n_] := With[{v = Table[Symbol["qb" <> ToString[i]], {i, 1, n}]},
    Total[Table[v[[i]]^2, {i, 1, n}]]];
qbVar[n_] := Table[Symbol["qb" <> ToString[i]], {i, 1, n}];
qbBox[n_] := And @@ Table[-5 <= Symbol["qb" <> ToString[i]] <= 5, {i, 1, n}];

ackObj[n_] := With[{v = Table[Symbol["ak" <> ToString[i]], {i, 1, n}]},
    -20 Exp[-0.2 Sqrt[Total[Table[v[[i]]^2, {i, 1, n}]]/n]]
    - Exp[Total[Table[Cos[2 Pi v[[i]]], {i, 1, n}]]/n] + 20 + E];
ackVar[n_] := Table[Symbol["ak" <> ToString[i]], {i, 1, n}];
ackBox[n_, w_] := And @@ Table[-w <= Symbol["ak" <> ToString[i]] <= w, {i, 1, n}];

katObj[n_] := With[{v = Table[Symbol["kt" <> ToString[i]], {i, 1, n}]},
    (10/n^2) Product[1 + i Sum[Abs[2^k v[[i]] - Round[2^k v[[i]]]]/2^k, {k, 1, 25}], {i, 1, n}]^(10/n^1.2) - (10/n^2)];
katVar[n_] := Table[Symbol["kt" <> ToString[i]], {i, 1, n}];
katBox[n_] := And @@ Table[-100 <= Symbol["kt" <> ToString[i]] <= 100, {i, 1, n}];

rasObj[n_] := With[{v = Table[Symbol["rs" <> ToString[i]], {i, 1, n}]},
    10 n + Total[Table[v[[i]]^2 - 10 Cos[2 Pi v[[i]]], {i, 1, n}]]];
rasVar[n_] := Table[Symbol["rs" <> ToString[i]], {i, 1, n}];
rasBox[n_] := And @@ Table[-5.12 <= Symbol["rs" <> ToString[i]] <= 5.12, {i, 1, n}];

rosObj[n_] := With[{v = Table[Symbol["ro" <> ToString[i]], {i, 1, n}]},
    Total[Table[100 (v[[i + 1]] - v[[i]]^2)^2 + (1 - v[[i]])^2, {i, 1, n - 1}]]];
rosVar[n_] := Table[Symbol["ro" <> ToString[i]], {i, 1, n}];
rosBox[n_] := And @@ Table[-5 <= Symbol["ro" <> ToString[i]] <= 5, {i, 1, n}];

maObj[n_] := With[{v = Table[Symbol["ma" <> ToString[i]], {i, 1, n}]},
    -Exp[-0.2 Sqrt[Total[Table[v[[i]]^2, {i, 1, n}]]/n]] Product[Cos[20 v[[i]]], {i, 1, n}]
    + 0.05 Total[Table[v[[i]]^2, {i, 1, n}]]];
maVar[n_] := Table[Symbol["ma" <> ToString[i]], {i, 1, n}];
maBox[n_] := And @@ Table[-5 <= Symbol["ma" <> ToString[i]] <= 5, {i, 1, n}];

gwObj[n_] := With[{v = Table[Symbol["gw" <> ToString[i]], {i, 1, n}]},
    1 - Exp[-2 Total[Table[(v[[i]] - 1.2345)^2, {i, 1, n}]]] + 10^-5 Total[Table[v[[i]]^2, {i, 1, n}]]];
gwVar[n_] := Table[Symbol["gw" <> ToString[i]], {i, 1, n}];
gwBox[n_] := And @@ Table[-5 <= Symbol["gw" <> ToString[i]] <= 5, {i, 1, n}];

pGri5 = griObj[5];   pGri5v = griVar[5];   pGri5b = griBox[5, 15];
pGri10 = griObj[10]; pGri10v = griVar[10]; pGri10b = griBox[10, 600];
pSq10 = sqObj[10];   pSq10v = sqVar[10];   pSq10b = sqBox[10];
pQb3 = qbObj[3];     pQb3v = qbVar[3];     pQb3b = qbBox[3];
pAck3 = ackObj[3];   pAck3v = ackVar[3];   pAck3b = ackBox[3, 32];
pKat8 = katObj[8];   pKat8v = katVar[8];   pKat8b = katBox[8];
pRas5 = rasObj[5];   pRas5v = rasVar[5];   pRas5b = rasBox[5];
pRas8 = rasObj[8];   pRas8v = rasVar[8];   pRas8b = rasBox[8];
pRos10 = rosObj[10]; pRos10v = rosVar[10]; pRos10b = rosBox[10];
pMa10 = maObj[10];   pMa10v = maVar[10];   pMa10b = maBox[10];
pGw10 = gwObj[10];   pGw10v = gwVar[10];   pGw10b = gwBox[10];

(* --- 1-D / 2-D, bare variables (compiled machine path) --- *)
bench["01 Chained trig 1D", NMinimize[{Sin[2 x] + Cos[x], -2 <= x <= 3}, {x}, Method -> R];];
check["01 Chained trig 1D", Round[10^4 First[NMinimize[{Sin[2 x] + Cos[x], -2 <= x <= 3}, {x}, Method -> R]]]];

bench["02 Quartic 1D", NMinimize[{x^4 - 3 x^2 - x, -5 <= x <= 5}, {x}, Method -> R];];
check["02 Quartic 1D", Round[10^4 First[NMinimize[{x^4 - 3 x^2 - x, -5 <= x <= 5}, {x}, Method -> R]]]];

bench["03 Gamma 1D", NMinimize[{Gamma[x], 1 <= x <= 2}, {x}, Method -> R];];
check["03 Gamma 1D", Round[10^5 First[NMinimize[{Gamma[x], 1 <= x <= 2}, {x}, Method -> R]]]];

bench["04 Schwefel 2D", NMinimize[{837.9658 - x Sin[Sqrt[Abs[x]]] - y Sin[Sqrt[Abs[y]]], -500 <= x <= 500 && -500 <= y <= 500}, {x, y}, Method -> R];];
check["04 Schwefel 2D", Round[10^3 First[NMinimize[{837.9658 - x Sin[Sqrt[Abs[x]]] - y Sin[Sqrt[Abs[y]]], -500 <= x <= 500 && -500 <= y <= 500}, {x, y}, Method -> R]]]];

bench["05 Schaffer N2 2D", NMinimize[{0.5 + (Sin[x^2 - y^2]^2 - 0.5)/(1 + 0.001 (x^2 + y^2))^2, -100 <= x <= 100 && -100 <= y <= 100}, {x, y}, Method -> R];];
check["05 Schaffer N2 2D", Round[10^4 First[NMinimize[{0.5 + (Sin[x^2 - y^2]^2 - 0.5)/(1 + 0.001 (x^2 + y^2))^2, -100 <= x <= 100 && -100 <= y <= 100}, {x, y}, Method -> R]]]];

bench["06 Rugged sine 2D", NMinimize[{x^2 + y^2 + 10 Sin[3 x]^2 + 10 Sin[3 y]^2, -5 <= x <= 5 && -5 <= y <= 5}, {x, y}, Method -> R];];
check["06 Rugged sine 2D", Round[10^4 First[NMinimize[{x^2 + y^2 + 10 Sin[3 x]^2 + 10 Sin[3 y]^2, -5 <= x <= 5 && -5 <= y <= 5}, {x, y}, Method -> R]]]];

bench["07 Eggholder 2D LF/20000", NMinimize[{-(y + 47) Sin[Sqrt[Abs[y + x/2 + 47]]] - x Sin[Sqrt[Abs[x - (y + 47)]]], -512 <= x <= 512 && -512 <= y <= 512}, {x, y}, Method -> Regg];];
check["07 Eggholder 2D LF/20000", Round[10^3 First[NMinimize[{-(y + 47) Sin[Sqrt[Abs[y + x/2 + 47]]] - x Sin[Sqrt[Abs[x - (y + 47)]]], -512 <= x <= 512 && -512 <= y <= 512}, {x, y}, Method -> Regg]]]];

(* --- 3-D..10-D, fresh-symbol objectives --- *)
bench["08 Griewank 5D", NMinimize[{Evaluate[pGri5], Evaluate[pGri5b]}, Evaluate[pGri5v], Method -> R];];
check["08 Griewank 5D", Round[10^4 First[NMinimize[{Evaluate[pGri5], Evaluate[pGri5b]}, Evaluate[pGri5v], Method -> R]]]];

bench["09 Griewank 10D", NMinimize[{Evaluate[pGri10], Evaluate[pGri10b]}, Evaluate[pGri10v], Method -> R];];
check["09 Griewank 10D", Round[10^4 First[NMinimize[{Evaluate[pGri10], Evaluate[pGri10b]}, Evaluate[pGri10v], Method -> R]]]];

bench["10 Scaled-quadratic 10D", NMinimize[{Evaluate[pSq10], Evaluate[pSq10b]}, Evaluate[pSq10v], Method -> R];];
check["10 Scaled-quadratic 10D", Round[10^4 First[NMinimize[{Evaluate[pSq10], Evaluate[pSq10b]}, Evaluate[pSq10v], Method -> R]]]];

bench["11 Quadratic bowl 3D", NMinimize[{Evaluate[pQb3], Evaluate[pQb3b]}, Evaluate[pQb3v], Method -> R];];
check["11 Quadratic bowl 3D", Round[10^4 First[NMinimize[{Evaluate[pQb3], Evaluate[pQb3b]}, Evaluate[pQb3v], Method -> R]]]];

bench["12 Ackley 3D", NMinimize[{Evaluate[pAck3], Evaluate[pAck3b]}, Evaluate[pAck3v], Method -> R];];
check["12 Ackley 3D", Round[10^4 First[NMinimize[{Evaluate[pAck3], Evaluate[pAck3b]}, Evaluate[pAck3v], Method -> R]]]];

bench["13 Katsuura 8D", NMinimize[{Evaluate[pKat8], Evaluate[pKat8b]}, Evaluate[pKat8v], Method -> R];];
check["13 Katsuura 8D", Round[10^4 First[NMinimize[{Evaluate[pKat8], Evaluate[pKat8b]}, Evaluate[pKat8v], Method -> R]]]];

bench["14 Rastrigin 5D", NMinimize[{Evaluate[pRas5], Evaluate[pRas5b]}, Evaluate[pRas5v], Method -> R];];
check["14 Rastrigin 5D", Round[10^4 First[NMinimize[{Evaluate[pRas5], Evaluate[pRas5b]}, Evaluate[pRas5v], Method -> R]]]];

bench["15 Rastrigin 8D", NMinimize[{Evaluate[pRas8], Evaluate[pRas8b]}, Evaluate[pRas8v], Method -> R];];
check["15 Rastrigin 8D", Round[10^4 First[NMinimize[{Evaluate[pRas8], Evaluate[pRas8b]}, Evaluate[pRas8v], Method -> R]]]];

bench["16 Rosenbrock 10D", NMinimize[{Evaluate[pRos10], Evaluate[pRos10b]}, Evaluate[pRos10v], Method -> R];];
check["16 Rosenbrock 10D", Round[10^3 First[NMinimize[{Evaluate[pRos10], Evaluate[pRos10b]}, Evaluate[pRos10v], Method -> R]]]];

bench["17 Modified Ackley 10D", NMinimize[{Evaluate[pMa10], Evaluate[pMa10b]}, Evaluate[pMa10v], Method -> R];];
check["17 Modified Ackley 10D", Round[10^4 First[NMinimize[{Evaluate[pMa10], Evaluate[pMa10b]}, Evaluate[pMa10v], Method -> R]]]];

bench["18 Gaussian well 10D LF/40000", NMinimize[{Evaluate[pGw10], Evaluate[pGw10b]}, Evaluate[pGw10v], Method -> Rgw];];
check["18 Gaussian well 10D LF/40000", Round[10^3 First[NMinimize[{Evaluate[pGw10], Evaluate[pGw10b]}, Evaluate[pGw10v], Method -> Rgw]]]];

File: benchmarks/84-direct-testbed/direct_testbed.py
#!/usr/bin/env python3
"""Experiment 84 -- DIRECT across the NMinimize test corpus (scipy column).

Same 18 problems as ``direct_testbed.m``, same labels, same order, solved with
scipy.optimize.direct on the identical bounded box. DIRECT is deterministic --
no seed. This is the RACEABLE subset of the NMinimize test corpus
(tests/test_nminimize.c): the continuous, box-bounded problems where the two
DIRECT implementations do the same work. Both columns use scipy's default
parameters except where a case is annotated ``LF/N`` (locally_biased=False,
maxfun=N) -- the same appropriate parameterisation on both sides.

The race is raw-vs-raw (scipy's direct does no local polish, so the .m column
sets "PostProcess" -> False). Every case reaches the SAME point in both systems:
the global for most, a shared non-global basin for Rosenbrock-10D (~8.7916) and
Gaussian-well-10D (both raw DIRECTs stall at 1.0 -- the well at 1.2345 is too
narrow for center-based subdivision).

The rest of the corpus is analysed in README.md and is NOT a DIRECT-vs-direct
race: scipy.optimize.direct is BOX-ONLY, so the constrained / equality /
disjunctive / integer problems have no scipy column (Mathilda's DIRECT solves
the low-dimensional ones via its penalty + augmented-Lagrangian polish, which
scipy's direct cannot do at all); the high-dimensional equality-constrained
A-series and the combinatorial B-series are beyond DIRECT in either system.

The check is the objective at the reported point, rounded per problem so both
implementations agree at the basin.
"""

import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import numpy as np
from scipy.optimize import direct
import scipy.special as _sp

from harness import bench, check, require

require(["scipy.optimize:direct"])


def _solve(fun, bounds, **kw):
    return direct(fun, bounds, **kw)


def _val(res, scale):
    return int(np.floor(scale * res.fun + 0.5))


# ---- objectives ----------------------------------------------------------

def _chained(v):    x = v[0]; return np.sin(2 * x) + np.cos(x)
def _gamma(v):      return float(_sp.gamma(v[0]))
def _quartic(v):    x = v[0]; return x ** 4 - 3 * x ** 2 - x

def _schwefel2(v):
    x, y = v
    return 837.9658 - x * np.sin(np.sqrt(abs(x))) - y * np.sin(np.sqrt(abs(y)))

def _schaffer2(v):
    x, y = v
    return 0.5 + (np.sin(x * x - y * y) ** 2 - 0.5) / (1 + 0.001 * (x * x + y * y)) ** 2

def _rugged(v):
    x, y = v
    return x * x + y * y + 10 * np.sin(3 * x) ** 2 + 10 * np.sin(3 * y) ** 2

def _eggholder(v):
    x, y = v
    return (-(y + 47) * np.sin(np.sqrt(abs(y + x / 2 + 47)))
            - x * np.sin(np.sqrt(abs(x - (y + 47)))))

def _griewank(v):
    v = np.asarray(v)
    return (np.sum(v ** 2 / 4000.0)
            - np.prod(np.cos(v / np.sqrt(np.arange(1, v.size + 1)))) + 1.0)

def _scaledquad(v):
    return float(sum(10.0 ** (6 * i / 9.0) * v[i] ** 2 for i in range(len(v))))

def _quadbowl(v):
    v = np.asarray(v); return float(np.sum(v * v))

def _ackley(v):
    v = np.asarray(v); n = v.size
    return (-20.0 * np.exp(-0.2 * np.sqrt(np.sum(v * v) / n))
            - np.exp(np.sum(np.cos(2 * np.pi * v)) / n) + 20.0 + np.e)

def _katsuura(v):
    v = np.asarray(v); d = v.size
    prod = 1.0
    for i in range(d):
        s = 1.0 + (i + 1) * sum(abs(2 ** k * v[i] - round(2 ** k * v[i])) / 2 ** k
                                for k in range(1, 26))
        prod *= s ** (10.0 / d ** 1.2)
    return (10.0 / d ** 2) * prod - (10.0 / d ** 2)

def _rastrigin(v):
    v = np.asarray(v)
    return 10.0 * v.size + np.sum(v * v - 10.0 * np.cos(2 * np.pi * v))

def _rosenbrock(v):
    v = np.asarray(v)
    return float(np.sum(100 * (v[1:] - v[:-1] ** 2) ** 2 + (1 - v[:-1]) ** 2))

def _modackley(v):
    v = np.asarray(v); d = v.size
    return (-np.exp(-0.2 * np.sqrt(np.sum(v * v) / d)) * np.prod(np.cos(20 * v))
            + 0.05 * np.sum(v * v))

def _gwell(v):
    v = np.asarray(v)
    return 1.0 - np.exp(-2.0 * np.sum((v - 1.2345) ** 2)) + 1e-5 * np.sum(v ** 2)


# ---- cases ---------------------------------------------------------------

_c01 = lambda: _solve(_chained,    [(-2, 3)])
_c02 = lambda: _solve(_quartic,    [(-5, 5)])
_c03 = lambda: _solve(_gamma,      [(1, 2)])
_c04 = lambda: _solve(_schwefel2,  [(-500, 500)] * 2)
_c05 = lambda: _solve(_schaffer2,  [(-100, 100)] * 2)
_c06 = lambda: _solve(_rugged,     [(-5, 5)] * 2)
_c07 = lambda: _solve(_eggholder,  [(-512, 512)] * 2, locally_biased=False, maxfun=20000)
_c08 = lambda: _solve(_griewank,   [(-15, 15)] * 5)
_c09 = lambda: _solve(_griewank,   [(-600, 600)] * 10)
_c10 = lambda: _solve(_scaledquad, [(-10, 10)] * 10)
_c11 = lambda: _solve(_quadbowl,   [(-5, 5)] * 3)
_c12 = lambda: _solve(_ackley,     [(-32, 32)] * 3)
_c13 = lambda: _solve(_katsuura,   [(-100, 100)] * 8)
_c14 = lambda: _solve(_rastrigin,  [(-5.12, 5.12)] * 5)
_c15 = lambda: _solve(_rastrigin,  [(-5.12, 5.12)] * 8)
_c16 = lambda: _solve(_rosenbrock, [(-5, 5)] * 10)
_c17 = lambda: _solve(_modackley,  [(-5, 5)] * 10)
_c18 = lambda: _solve(_gwell,      [(-5, 5)] * 10, locally_biased=False, maxfun=40000)

bench("01 Chained trig 1D", _c01);         check("01 Chained trig 1D", _val(_c01(), 1e4))
bench("02 Quartic 1D", _c02);              check("02 Quartic 1D", _val(_c02(), 1e4))
bench("03 Gamma 1D", _c03);                check("03 Gamma 1D", _val(_c03(), 1e5))
bench("04 Schwefel 2D", _c04);             check("04 Schwefel 2D", _val(_c04(), 1e3))
bench("05 Schaffer N2 2D", _c05);          check("05 Schaffer N2 2D", _val(_c05(), 1e4))
bench("06 Rugged sine 2D", _c06);          check("06 Rugged sine 2D", _val(_c06(), 1e4))
bench("07 Eggholder 2D LF/20000", _c07);   check("07 Eggholder 2D LF/20000", _val(_c07(), 1e3))
bench("08 Griewank 5D", _c08);             check("08 Griewank 5D", _val(_c08(), 1e4))
bench("09 Griewank 10D", _c09);            check("09 Griewank 10D", _val(_c09(), 1e4))
bench("10 Scaled-quadratic 10D", _c10);    check("10 Scaled-quadratic 10D", _val(_c10(), 1e4))
bench("11 Quadratic bowl 3D", _c11);       check("11 Quadratic bowl 3D", _val(_c11(), 1e4))
bench("12 Ackley 3D", _c12);               check("12 Ackley 3D", _val(_c12(), 1e4))
bench("13 Katsuura 8D", _c13);             check("13 Katsuura 8D", _val(_c13(), 1e4))
bench("14 Rastrigin 5D", _c14);            check("14 Rastrigin 5D", _val(_c14(), 1e4))
bench("15 Rastrigin 8D", _c15);            check("15 Rastrigin 8D", _val(_c15(), 1e4))
bench("16 Rosenbrock 10D", _c16);          check("16 Rosenbrock 10D", _val(_c16(), 1e3))
bench("17 Modified Ackley 10D", _c17);     check("17 Modified Ackley 10D", _val(_c17(), 1e4))
bench("18 Gaussian well 10D LF/40000", _c18); check("18 Gaussian well 10D LF/40000", _val(_c18(), 1e3))
